File: 9_carFuel/carFuelCpp/car_fueling.cpp
#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::vector;
using std::max;

int compute_min_refills(int dist, int tank, vector<int> & stops) {
    int numRefills = 0;
    int curRefill = 0;
    int lastRefill = 0;
    
    int n = stops.size();
    // cout << n << "\n";
    
    std::vector<int>::iterator it;
    it = stops.begin();
    stops.insert(it, 0);
    stops.push_back(dist);
/*    
    for (int i = 0; i <= n; ++i) {
      cout << stops[i] << " ";
    }
    cout << "\n";
 */
    
    while (curRefill <= n) {
      lastRefill = curRefill;   
      while (curRefill <= n && 
            ((stops[curRefill + 1] - stops[lastRefill]) <= tank)) {
        curRefill = curRefill + 1;  
      } // inner while
      if (curRefill == lastRefill)
        return -1;
      if (curRefill <= n)
        numRefills = numRefills + 1;
    } // outer while
/*    
    cout << "cr = " << curRefill << "\n";
    cout << "lr = " << lastRefill << "\n";
 */
    
    return numRefills;
}

int main() {
    int d = 0;
    cin >> d;
    int m = 0;
    cin >> m;
    int n = 0;
    cin >> n;

    vector<int> stops(n);
    for (size_t i = 0; i < n; ++i)
        cin >> stops.at(i);

    cout << compute_min_refills(d, m, stops) << "\n";

    return 0;
}

File: 3-4_carFuel/carFuelJava/CarFueling.java
import java.util.*;
import java.io.*;

public class CarFueling {
	
    static int computeMinRefills(int dist, int tank, int[] stops) {
        int numRefills = 0;
        int curRefill = 0;
        int lastRefill = 0;
        
        // Add begin and end pt
        int n = stops.length + 2;        
        int[] stopsAll = new int[n];
        
        // Add beginning and end to new array
        stopsAll[0] = 0;        
        for (int i = 1; i < n - 1; i++)
        	stopsAll[i] = stops[i - 1];        
        stopsAll[n - 1] = dist;
        
        // debug: stopsAll array
        // for (int i = 0; i < n; i++)
        //	  System.out.print(stopsAll[i] + " ");
        // System.out.println();
        
        
        while (curRefill < n - 1) {
        	lastRefill = curRefill;        	
        	while (curRefill < n - 1 && 
                    ((stopsAll[curRefill + 1] - stopsAll[lastRefill]) <= tank)) {
                curRefill = curRefill + 1;  
            } // inner while
 
        	if (curRefill == lastRefill)
                return -1;
            if (curRefill < n - 1)
                numRefills = numRefills + 1;
        } // outer while
        
        
        
        return numRefills;
    } // end computeMinRefills

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int dist = scanner.nextInt();
        int tank = scanner.nextInt();
        int n = scanner.nextInt();
        int stops[] = new int[n];
        for (int i = 0; i < n; i++) {
            stops[i] = scanner.nextInt();
        }

        System.out.println(computeMinRefills(dist, tank, stops));
        scanner.close();
    }
}
